### ISSUE TITLE
Beiwe server issues: Watch_processing() stuck at 120564

### ISSUE THREAD

**Issue by justintbaker** [onnela-lab/beiwe-backend#284]
Hi @biblicabeebli @zagorsky -- hoping one of you may know how to address my current issue.

Our Beiwe Instance had been running smoothly for several months, until it crashed about 1 month ago, which in retrospect was due to running out of storage on the modestly sized (8Gb) volume attached to the data processing manager.  (With @zagorsky 's help, I had already expanded the space on the main EC2 instance machine.).

After restarting the instance and expanding the available space on the data processing manager machine,  >120K files were unprocessed, which is a lot more than we have encountered previously, partly due to the longer than usual time before I was made aware of the issue.  

This might have been okay had the system started chugging through these -- however, it appears stuck: exactly 120564 files for the past 3-4d.   When I use watch_processing() on the data processing manager, I just see lines like this over and over:

`=================================

2022-03-14 16:03:47.289251+00:00: 120654 files to process
2022-03-14 16:03:47.341631+00:00 active tasks: [292, 77, 300, 298, 294, 296, 293, 290, 297, 299]
2022-03-14 16:03:47.457502+00:00 scheduled tasks: []
2022-03-14 16:03:47.567182+00:00 registered tasks: [117, 156, 86, 141, 163, 232, 88, 119, 142, 145, 190, 92, 120, 237, 193, 218, 98, 121, 146, 106, 125, 149, 150, 233, 111, 126, 252, 151, 113, 139, 116, 152]

=================================`

Also, the load on the data processing manager is consistently >10 as it attempts to spin up a million python jobs like this one:
 
/home/ubuntu/.pyenv/versions/beiwe/bin/python -m celery -A services.celery_data_processing worker -Q data_processing --loglevel=info -Ofair --hostname=%h_processing --autoscale=10,2
 
But I have a feeling they are not completing or the # of files to process would be coming down, however gradually.

I’ve tried restarting the manager a couple times and redeployed the server, but the behavior is the same.  Any chance anyone has suggestions for how to kick the system into at least processing this backlog?  I could maybe expand the instance size (currently T3.medium) or increase number of EC2 instances. 
  
Thanks for any assistance you may be able to provide!!

**Comment by biblicabeebli** [comment]
Based on the presence of --autoscale=10,2 you appear to be relatively up to date, but if not do that first.  There are recent data processing optimizations.

This sounds like an out-of-memory issue, when data processing slows down that means there is a lot of paging occurring on the swap file.  You can try an M- C- or T- series servers, anthing with more ram.

You can also try removing --autoscale=10,2 , that will reduce the number of concurrent processes.

**Comment by justintbaker** [comment]
Thanks @biblicabeebli  -- Yes I recently redeployed off the "main" branch.  

Can you point me to how to change the type of EC2 instance that gets launched for the data processing server?  Is this within the server_configuration.txt so I can change on my local pre-deployment machine?  Or would I be able to change this within the AWS Console in a way that would persist?

Similarly for "removing --autoscale=10,2" -- would I achieve this on the data processing server itself by altering celery?  From I can tell in "htop", the process coordinating everything (just below supervisord) is:

`/home/ubuntu/.pyenv/versions/beiwe/bin/python -m celery -A services.celery_push_notifications worker -Q push_notifications --loglevel=info -Ofair --hostname=%h_notifications --concurrency=20 --pool=threads`  

Is there a way to manually run this or tell supervisord to run with different arguments?

Also, within the celery_processing.log file, 
[beiwe_celery_processing-22-03-16.txt](https://github.com/onnela-lab/beiwe-backend/files/8265944/beiwe_celery_processing-22-03-16.txt) -- it appears files are being processed, but here as well, the file numbers on a participant basis are not decreasing.  The same ID shows up ~6 sec with the same number of files to process.

**Comment by justintbaker** [comment]
@biblicabeebli @zagorsky One additional piece of information.  I spent a couple hours today with an AWS consultant, who helped me troubleshoot a few options:

1) We boosted the size of the data processing server to r6i.4xlarge (128Gb Memory), but this does not prevent jobs from dying.

2) We also modified "install_celery_worker.sh" script to change concurrency=1 (for `celery_push_send`)

`command = /home/ubuntu/.pyenv/versions/beiwe/bin/python -m celery -A services.celery_push_notifications worker -Q push_notifications --loglevel=info -Ofair --hostname=%%h_notifications --concurrency=1 --pool=threads`

...and removed the autoscale from `celery_processing`:

`command = /home/ubuntu/.pyenv/versions/beiwe/bin/python -m celery -A services.celery_forest worker -Q forest_queue --loglevel=info -Ofair --hostname=%%h_forest`

These changes did not fix the issue, but we started seeing many many lines saying:

`[2022-03-23 19:36:21,467: WARNING/ForkPoolWorker-22] Connection pool is full, discarding connection: beiwe-data-beiweitp-clo4quebjno9r2dtente6wbsmisvjglizgqp9lmoo8y.s3.amazonaws.com`

Of note, when I SSH into the main Beiwe EC2 instance (using "eb ssh beiweitp"), I am not able to run "db":

[ec2-user@ip-172-31-19-204 ~]$ db
-bash: db: command not found

This makes me wonder whether the instance is just not deployed correctly, so there is nothing for the data processing server to connect to.  I have attempted to "eb deploy beiweitp" multiple times using the latest beiwe-backend "main" branch, and this seems to work with no errors.

**Comment by biblicabeebli** [comment]
Hi Justin, it took me a while to get caught up and back to this, I'll try to address everything relevant here.


After looking at that log file I don't think you need to run on a larger server - this makes sense, I've worked hard to reduce that overhead - I explain below.


Specifying server size for data processing servers:
server sizes are declared in the `cluster_management/environment_configuration/{your EB environment name here}_server_settings.json`. Just change the values of the


To change the supervisord configuration on the fly:
There is an alias `superconf` on the processing servers that will open the supervisord configuration file in nano for you.  From there you can edit the commands.  To apply those you need to then run the `processing-stop` alias, and then the `processing-start` alias after waiting for `supervisord` to exit.  You can watch for that using `htop`.

From your log file, this line:
```
Error running data processing: column database_summarystatisticdaily.beiwe_accelerometer_bytes does not exist
LINE 1: ...ant_id", "database_summarystatisticdaily"."date", "database_...
```
That means you have a database schema mismatch between those declared in the python code and the tables in the database.  You will need to deploy to your Elastic Beanstalk servers to run the database migration, orrr you can yolo it and run `python manage.py migrate` from the data processing server.  Note that doing that is quite plausibly going to screw up ... random unpredictable stuff on your EB servers due to potential schema mismatches.

Those customizations you made on the celery/supervisor config:
You should revert to the original commands in the celery configuration and just remove the autoscale option.  The default concurrency is the number of cpu cores, and I think `concurrency=1` with `pool=threads` is an illegal combination because that only allows for the dispatch thread with no consumer threads... maybe.  (Frankly, Celery's errors on these basic configuration options are deeply incoherent and frequently outright misleading.)

Those deploys not working and the missing db aliases:
Um... I have no idea... Sorry.
Make sure you are up to date on the codebase... which you say you are.
If it says it succeeds that means the new code is live - potentially with a 1-2 minute lag time on the server literally restarting Apache - but all the configuration steps are declared in the `.ebextensions/01.config` file, including placing the bash profile file that contains the declaration of the db alias.  So if its not there then the deploy failed and the `eb` tool will definitely tell you this.
Any chance you are deploying to the wrong EB environment somehow?  The python 3.6 platform-compatible code (branch `python-3.6`) is fundamentally incompatible with a successful `eb deploy` operation on a python 3.8 platform, the reverse is truuue with any commit after the `pythos-3.8-redeploy` branch being incompatible with a python 3.6 platform EB server.

Try an `eb list` to see your different eb environments, maybe you got some commands crossed and are pointing at the wrong environments after the 3.8 platform update, or maybe you never changed your urls so it has been super confusing because 3.6 is still up?

**Comment by justintbaker** [comment]
Thanks @biblicabeebli -- I've tried a few things; haven't made a ton of progress.  Here's the status:

1. I've reverted the `celery` customizations aside from removing the `autoscale` option.

2. I checked `eb list` but alas it just the one `beiweitp` instance.

3. Regarding "deploy your EB servers to run the database migration" -- this seems important, since from what I can tell nothing useful is running on the actual `beiweitp` EC2 instance.  

I have tried redeploying several times, and there is no indication that the `eb deploy beiweitp` fails.  When I launch a data processing server, I have confirmed that it runs python 3.8, and I can see when creating the processing manager that it is replacing bash_profile, etc.  However, when I check python version on the main `beiweitp` instance I am seeing `2.7.18`, and other things like `htop` are missing completely.  

Here's the output `eb_status` which hopefully provides clues...there the python version says 3.6: 

`Environment details for: beiweitp
  Application name: beiwe-application
  Region: us-east-1
  Deployed Version: app-220330_155746
  Environment ID: e-zxuzse3ntn
  Platform: arn:aws:elasticbeanstalk:us-east-1::platform/Python 3.6 running on 64bit Amazon Linux/2.10.11
  Tier: WebServer-Standard-1.0
  CNAME: beiweitp.d4km3wazh7.us-east-1.elasticbeanstalk.com
  Updated: 2022-04-03 19:44:33.805000+00:00
  Status: Ready
  Health: Grey
Alert: Your environment is using a deprecated platform branch. It might not be supported in the future.`
